### PR TITLE
feat: Extension inference for CFGs

### LINKED ISSUE #497: Extension inference for control flow graphs
Blocks #473

### LINKED ISSUE #526: Extension inference for CFGs

### PR DESCRIPTION
**Pull request by croyzor**
Based off of #527. Add constraints for control flow graphs and their basic blocks, and a test of a CFG which adds resources in its basic blocks.
Resolves #497

### CODE REVIEW

File: src/extension/infer.rs [hunk 1]
```diff
@@ -302,6 +302,18 @@ impl UnificationContext {
...
             }
 
+            if node_type.tag() == OpTag::Cfg {
+                let mut children = hugr.children(node);
+                let entry = children.next().unwrap();
+                let exit = children.next().unwrap();
+                let m_entry = self.make_or_get_meta(entry, Direction::Incoming);
+                let m_exit_in = self.make_or_get_meta(exit, Direction::Incoming);
+                let m_exit_out = self.make_or_get_meta(exit, Direction::Outgoing);
+                self.add_constraint(m_input, Constraint::Equal(m_entry));
```
Comment (acl-cqc) [inline src/extension/infer.rs:312]: It seems a bit odd that you deal with both Direction::Incoming and Direction::Outgoing for the exit node, but not for the entry node. Do you mean to do similarly for `self.make_or_get_meta(entry, Direction::Outgoing)` too?

If so, how about commoning up with the case for nested dataflow nodes above, i.e. breaking this block out into a helper function:
```
for dir in Direction::BOTH {
                    let m_input_node = self.make_or_get_meta(input, dir);
                    self.add_constraint(m_input_node, Constraint::Equal(m_input));
                    let m_output_node = self.make_or_get_meta(output, dir);
                    self.add_constraint(m_output_node, Constraint::Equal(m_output));
                }
```

File: src/extension/infer.rs [hunk 2]
```diff
@@ -1092,4 +1104,183 @@ mod test {
...
+        let a = ExtensionSet::singleton(&A);
+        let abc = ExtensionSet::from_iter([A, B, C]);
+        let bc = ExtensionSet::from_iter([B, C]);
+        let b = ExtensionSet::singleton(&B);
+        let c = ExtensionSet::singleton(&C);
+
+        let oneway = vec![Type::new_predicate([type_row![NAT]])];
+        let twoway = vec![Type::new_predicate([type_row![NAT], type_row![NAT]])];
+
+        let mut hugr = Hugr::new(NodeType::open_extensions(ops::CFG {
```
Comment (acl-cqc) [inline src/extension/infer.rs:1147]: driveby, but we should probably use FunctionType here (inside CFG), too

File: src/extension/infer.rs (hunk 1 above)
Comment (croyzor) [reply src/extension/infer.rs:312]: The exit node of a CFG can never add extension requirements between its inputs and outputs, but an entry node is an arbitrary DFB with an extension_delta, so the output extensions aren't necessarily the same as the CFG's input extensions.

Having said that, we _can_ add a constraint that the entry node's outputs are the CFG input plus that node's delta

File: src/extension/infer.rs [hunk 3]
```diff
@@ -1092,4 +1104,183 @@ mod test {
...
+    ///          |  (Adds [A]) |
+    ///          +-/---------\-+
+    ///           /           \
+    ///  +-------/-----+     +-\-------------+
+    ///  |     BB0     |     |      BB1      |
+    ///  | (Adds [BC]) |     |   (Adds [B])  |
+    ///  +----\--------+     +---/------\----+
+    ///        \                /        \
+    ///         \              /          \
+    ///          \       +----/-------+  +-\---------+
```
Comment (acl-cqc) [inline src/extension/infer.rs:1128]: I think you could make this a slightly more interesting test by linking BB0 to BB10 (rather than to exit)....but,

I think you could make this a *lot* more interesting by linking, say, `BB10`, *back* to `BB1` or, even better, `Entry` to make a loop. I guess the hope is that unification will deal with loops no problem, without requiring iteration to a fixpoint (say), but it'd be great to have the test of that :)

File: src/builder/cfg.rs
Comment (acl-cqc) [inline src/builder/cfg.rs:1]: I'd drop changes to this file, you've not got anything else here

File: src/extension/infer.rs [hunk 4]
```diff
@@ -1092,4 +1104,183 @@ mod test {
...
+
+        let mkpred = hugr.add_node_with_parent(
+            entry,
+            NodeType::open_extensions(make_opaque(
+                A,
+                FunctionType::new(vec![NAT], twoway.clone()).with_extension_delta(&a),
+            )),
+        )?;
+
+        let [bb0, bb0_in, bb0_out] = create_with_io(
```
Comment (acl-cqc) [inline src/extension/infer.rs:1181]: You've got four lots here of `create_with_io` + `add_node_with_parent` + 2*`connect` ("internal wiring"). I'd say that was worth making a helper function or lambda, e.g. inside this test. (It can probably create the BasicBlock::DFB as well.)

File: src/extension/infer.rs [hunk 5]
```diff
@@ -1092,4 +1104,183 @@ mod test {
...
+        hugr.connect(entry, 0, bb0, 0)?;
+        hugr.connect(entry, 1, bb1, 0)?;
+        hugr.connect(bb1, 0, bb10, 0)?;
+        hugr.connect(bb1, 1, bb11, 0)?;
+
+        hugr.connect(bb0, 0, exit, 0)?;
+        hugr.connect(bb10, 0, exit, 0)?;
+        hugr.connect(bb11, 0, exit, 0)?;
+
+        hugr.infer_extensions()?;
```
Comment (acl-cqc) [inline src/extension/infer.rs:1282]: The idea is that you've fixed them, so this also checks consistency with the extensions specified already?

**Review by acl-cqc — COMMENTED** [review]
Looks good, thanks - I presume you've already handled the case of ControlFlow edges?

**Comment by croyzor** [comment]
> I presume you've already handled the case of ControlFlow edges?

Yes, this is done as a drive-by in #527

File: src/extension/infer.rs (hunk 1 above)
Comment (acl-cqc) [reply src/extension/infer.rs:312]: Ah yes sorry, I forget that Entry could have anything in it. Hmmm. So that case is handled elsewhere. Does the op_signature for an Output node not specify that it adds no delta, then? (i.e. do we need to make a constraint for Exit node output extensions here at all?)

Comment (croyzor) [reply src/extension/infer.rs:312]: Yes you're right the output constraint is probably redundant

Comment (acl-cqc) [reply src/extension/infer.rs:312]: Ok in that case I suggest you do either *one* of the following two options:
* Don't add output constraint for either node. *Ideally* I think you'd make the same change to the `if let Some([input, output]) = hugr.get_io(node)` block above too, and perhaps elsewhere as a general principle, tho, and that might be better done in a later PR. In which case:
* Just add a comment here as to why entry and exit are different. (I should know, I wrote the spec for them, but having not worked on CFGs for awhile I forgot, and so others will too - it is an oddity of the spec that they are not symmetric, so I think that justifies a comment.)

File: src/algorithm/nest_cfgs.rs [hunk 6]
```diff
@@ -428,8 +428,8 @@ pub(crate) mod test {
         //               \-> right -/             \-<--<-/
         let mut cfg_builder = CFGBuilder::new(FunctionType::new(type_row![NAT], type_row![NAT]))?;
 
-        let pred_const = cfg_builder.add_constant(Const::simple_predicate(0, 2))?; // Nothing here cares which
-        let const_unit = cfg_builder.add_constant(Const::simple_unary_predicate())?;
+        let pred_const = cfg_builder.add_constant(Const::simple_predicate(0, 2), None)?; // Nothing here cares which
```
Comment (acl-cqc) [inline src/algorithm/nest_cfgs.rs:431]: Good to update the comment here - "Nothing here cares which *branch*" to be clear that the comment isn't about the extensions. (However, I think this should not actually be showing up in this PR, so this might disappear post-merge)

File: src/algorithm/nest_cfgs.rs [hunk 7]
```diff
@@ -675,8 +675,8 @@ pub(crate) mod test {
         separate: bool,
     ) -> Result<(Hugr, BasicBlockID, BasicBlockID), BuildError> {
         let mut cfg_builder = CFGBuilder::new(FunctionType::new(type_row![NAT], type_row![NAT]))?;
-        let pred_const = cfg_builder.add_constant(Const::simple_predicate(0, 2))?; // Nothing here cares which
-        let const_unit = cfg_builder.add_constant(Const::simple_unary_predicate())?;
+        let pred_const = cfg_builder.add_constant(Const::simple_predicate(0, 2), None)?; // Nothing here cares which
```
Comment (acl-cqc) [inline src/algorithm/nest_cfgs.rs:678]: ditto "Nothing here cares which branch"

File: src/builder/build_traits.rs [hunk 8]
```diff
@@ -62,14 +62,25 @@ pub trait Container {
...
     /// # Errors
     ///
     /// This function will return an error if there is an error in adding the
     /// [`OpType::Const`] node.
-    fn add_constant(&mut self, constant: ops::Const) -> Result<ConstID, BuildError> {
-        let const_n = self.add_child_op(constant)?;
+    fn add_constant(
+        &mut self,
+        constant: ops::Const,
+        provenance: Option<ExtensionId>,
```
Comment (acl-cqc) [inline src/builder/build_traits.rs:77]: Didn't we say 'provenance' should not be an Id but a Set? I'm not sure whether an Option<ExtensionSet> or just an ExtensionSet, depending on whether None means "no extensions" (=empty ExtensionSet) or "unknown extensions" (different from an empty ExtensionSet)...

Comment (acl-cqc) [reply src/builder/build_traits.rs:77]: Oh, again this might be the wrong PR

File: src/builder/build_traits.rs [hunk 9]
```diff
@@ -190,7 +201,7 @@ pub trait Dataflow: Container {
         op: impl Into<OpType>,
         input_wires: impl IntoIterator<Item = Wire>,
     ) -> Result<BuildHandle<DataflowOpID>, BuildError> {
-        self.add_dataflow_node(NodeType::pure(op), input_wires)
+        self.add_dataflow_node(NodeType::open_extensions(op), input_wires)
```
Comment (acl-cqc) [inline src/builder/build_traits.rs:204]: I'm gonna stop commenting about things like this that look a bit out of date until you've done some merging/rebasing/something....

File: src/extension/infer.rs [hunk 10]
```diff
@@ -323,10 +335,11 @@ impl UnificationContext {
         for tgt_node in hugr.nodes() {
             let sig: &OpType = hugr.get_nodetype(tgt_node).into();
             // Incoming ports with a dataflow edge
-            for port in hugr
-                .node_inputs(tgt_node)
-                .filter(|src_port| matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_))))
-            {
+            for port in hugr.node_inputs(tgt_node).filter(|src_port| {
+                matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_)))
```
Comment (acl-cqc) [inline src/extension/infer.rs:339]: Yes, you *can* combine these into a single `matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_)) | Some(EdgeKind::Static(_)) | Some(EdgeKind::ControlFlow))`, I checked...

File: src/extension/infer.rs (hunk 2 above)
Comment (acl-cqc) [reply src/extension/infer.rs:1147]: Really I think that would be better as a separate PR but I might be ok...

File: src/extension/infer.rs [hunk 11]
```diff
@@ -324,8 +334,12 @@ impl UnificationContext {
             let sig: &OpType = hugr.get_nodetype(tgt_node).into();
             // Incoming ports with a dataflow edge
             for port in hugr.node_inputs(tgt_node).filter(|src_port| {
-                matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_)))
-                    || matches!(sig.port_kind(*src_port), Some(EdgeKind::Static(_)))
+                matches!(
+                    sig.port_kind(*src_port),
+                    Some(EdgeKind::Value(_))
+                        | Some(EdgeKind::Static(_))
+                        | Some(EdgeKind::ControlFlow)
```
Comment (acl-cqc) [inline src/extension/infer.rs:341]: I think this neat little one-line change is doing most of the work!?

Which is cool. But can you
1. update the comment two lines above: "Incoming ports with a dataflow edge" doesn't seem right anymore.
2. (super-nit) 10 lines above, "Seperate loop so that...." -> "Separate loop so that". The former is an adjective, the latter a verb...

File: src/extension/infer.rs [hunk 12]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+        let opaque =
+            ops::custom::OpaqueOp::new(extension.into(), "", "".into(), vec![], Some(signature));
+        ops::custom::ExternalOp::from(opaque).into()
+    }
+
+    fn make_block(
+        hugr: &mut Hugr,
+        bb_parent: Node,
+        dfb: impl Into<OpType>,
+        dfb_sig: FunctionType,
```
Comment (acl-cqc) [inline src/extension/infer.rs:1177]: I note the following in `impl Case`: https://github.com/CQCL-DEV/hugr/blob/554d658f7afd38f0eb567ceff0599e65e74541c9/src/ops/controlflow.rs#L237-L238

You could add something similar in `controlflow.rs` for `impl BasicBlock` (I guess it would have to return an Option to allow for BasicBlock::Exit), and then drop the `dfg_sig` here, making `dfb` be an `impl Into<BasicBlock>`. (More thorough still would be to turn the contents of BasicBlock::DFB into a separate struct and impl for *that*, but I'd be happy not to go as far as that in this PR)

File: src/extension/infer.rs [hunk 13]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+                extension_delta: c.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                C,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&c),
+            ),
+        )?;
+
+        let bb11 = make_block(
```
Comment (acl-cqc) [inline src/extension/infer.rs:1309]:
```suggestion
        let bb11 = bb10.clone();
```

File: src/extension/infer.rs [hunk 14]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+                extension_delta: c.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                C,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&c),
+            ),
+        )?;
+
+        // Internal wiring for DFGs
```
Comment (acl-cqc) [inline src/extension/infer.rs:1325]: Move this up to where `entry`(_in,_out) and `mkpred` are declared

Comment (acl-cqc) [reply src/extension/infer.rs:1325]: (Like you are doing with `let entry =` in the `multi_entry` test below, tho I won't insist on hiding all these in a nested scope)

File: src/extension/infer.rs [hunk 15]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+        )?;
+        let entry = {
+            let dfb = ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: ExtensionSet::new(),
+            };
+
+            let entry = hugr.add_node_before(exit, NodeType::open_extensions(dfb))?;
```
Comment (acl-cqc) [inline src/extension/infer.rs:1364]: Is this not `let (entry, entry_in, entry_out) = create_with_io(.....)` ? (If you do `entry` before `exit` maybe)

File: src/extension/infer.rs [hunk 16]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: ExtensionSet::new(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(PRELUDE_ID, FunctionType::new(vec![NAT], oneway(NAT))),
+        )?;
+
+        let bb1 = make_block(
```
Comment (acl-cqc) [inline src/extension/infer.rs:1407]:
```suggestion
        let bb1 = bb0.clone()
```

File: src/extension/infer.rs [hunk 17]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: ExtensionSet::new(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(PRELUDE_ID, FunctionType::new(vec![NAT], oneway(NAT))),
+        )?;
+
+        // Mult
+        let bb2 = make_block(
```
Comment (acl-cqc) [inline src/extension/infer.rs:1421]:
```suggestion
        let bb2 = bb0.clone();
```

File: src/extension/infer.rs [hunk 18]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+        hugr.connect(bb10, 0, exit, 0)?;
+        hugr.connect(bb11, 0, exit, 0)?;
+
+        hugr.infer_extensions()?;
+
+        Ok(())
+    }
+
+    #[test]
+    fn multi_entry() -> Result<(), Box<dyn Error>> {
```
Comment (acl-cqc) [inline src/extension/infer.rs:1345]: No ascii art here, but this looks like entry -> (if(P) bb0; else bb1;) -> bb2 -> exit, which seems like a strictly simpler/easier case than the above. I guess that's OK, ascii art is nice tho (even if just a three-line comment), and why "multi_entry" ??

File: src/extension/infer.rs [hunk 19]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+            let dfb: OpType = ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: entry_ext.clone(),
+            }
+            .into();
+
+            let entry = hugr.add_node_before(exit, NodeType::open_extensions(dfb))?;
+            let entry_in = hugr.add_node_with_parent(
```
Comment (acl-cqc) [inline src/extension/infer.rs:1493]: again this looks like `let (entry, entry_in, entry_out) = create_with_io(.....)` ?

File: src/extension/infer.rs [hunk 20]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+        hugr.connect(bb2, 0, entry, 0)?;
+
+        Ok(hugr)
+    }
+
+    #[test]
+    fn test_cfg_loops() -> Result<(), Box<dyn Error>> {
+        let just_a = ExtensionSet::singleton(&A);
+        let variants = vec![
+            (
```
Comment (acl-cqc) [inline src/extension/infer.rs:1563]: You could add a couple more here - with `just_a` repeated, and one with two blocks that do something. I mean, since you've set up this nifty framework already, I mean :-)

File: src/extension/infer.rs (hunk 12 above)
Comment (acl-cqc) [reply src/extension/infer.rs:1177]: Or just make `make_block` here take a TypeRow for inputs and a Vec<TypeRow> for predicate, and do the rest here

File: src/extension/infer.rs [hunk 21]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: ExtensionSet::new(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(PRELUDE_ID, FunctionType::new(vec![NAT], oneway(NAT))),
+        )?;
+
+        // Mult
```
Comment (acl-cqc) [inline src/extension/infer.rs:1420]: ??

Comment (acl-cqc) [reply src/extension/infer.rs:1420]: Multiple predecessors?

File: src/extension/infer.rs [hunk 22]
```diff
@@ -1133,6 +1160,502 @@ mod test {
...
+        let root = hugr.root();
+
+        let exit = hugr.add_node_with_parent(
+            root,
+            NodeType::open_extensions(ops::BasicBlock::Exit {
+                cfg_outputs: type_row![NAT],
+            }),
+        )?;
+
+        let entry = {
```
Comment (acl-cqc) [inline src/extension/infer.rs:1604]: One more create_with_io ? Or, better, why not make `make_block` return all three blocks

**Review by acl-cqc — APPROVED** [review]
A few ideas here and there, but this is great, thank you - there's nothing unexpected, mostly things just work and look as it feels they should, and most of the PR is a great set of tests :-). (So most comments are on the tests!)

File: src/extension/infer.rs (hunk 13 above)
Comment (croyzor) [reply src/extension/infer.rs:1309]: Not quite, because `make_block` is adding a bunch of nodes to the graph, but I can certainly abstract this a bit!

File: src/extension/infer.rs (hunk 15 above)
Comment (croyzor) [reply src/extension/infer.rs:1364]: Hm, I thought about this but opted against it because defining entry before exit like that is taking advantage of implementation details and using `add_node_before` is more "the right thing"

File: src/extension/infer.rs (hunk 19 above)
Comment (croyzor) [reply src/extension/infer.rs:1493]: Here (and below) `create_with_io` is avoided because the dataflow parent needs to be added with `add_node_before` rather than just `add_node_with_parent`

File: src/extension/infer.rs (hunk 13 above)
Comment (acl-cqc) [reply src/extension/infer.rs:1309]: ah yes, good point, `clone`ing an op doesn't clone its children, sorry!

File: src/extension/infer.rs (hunk 15 above)
Comment (acl-cqc) [reply src/extension/infer.rs:1364]: Sorry, I am barking up the wrong tree here. This is nothing like `create_with_io` - that makes an Input and Output (dataflow) nodes, this makes an entry (BasicBlock::DFB) and exit (BasicBlock::Exit) node.

That said, given you do this a lot across three or four tests, maybe you could just have a helper `create_with_entry_exit`....
